tortoise/fixture_test.go

```diff
@@ -0,0 +1,791 @@
+		ballot.base = value.baseBallot
+		ballot.Opinion.Votes = value.votes
+	})
+	return b
+}
+
+func (a *atxAction) execute(trt *Tortoise) {
+	trt.OnAtx(&a.header)
+}
+
+func (a *atxAction) ballot(n int, opts ...*bopt) *ballotAction {
+	lid := uint32(n) + types.GetEffectiveGenesis().Uint32()
+	if val, exist := a.ballots[lid]; exist {
+		return val
+	}
+	b := types.BallotTortoiseData{}
+	b.AtxID = a.header.ID
+	b.Layer = types.LayerID(lid)
+	hs := hash.Sum(a.header.ID[:], []byte(strconv.Itoa(int(lid))))
+	copy(b.ID[:], hs[:])
+
+	val := &ballotAction{
+		BallotTortoiseData: b,
+	}
+	for _, opt := range opts {
+		for _, o := range opt.opts {
+			o(val)
+		}
+	}
+	if a.reference == nil {
+		a.reference = val
+	} else {
+		val.Ref = &a.reference.ID
+		val.before = append(val.before, a.reference)
+	}
+	if (val.Opinion.Hash == types.Hash32{}) {
+		val.Opinion.Hash = val.computeOpinion()
+	}
+
+	if val.base != nil {
+		val.before = append(val.before, val.base)
+	}
+	for _, atx := range val.activeset {
+		val.before = append(val.before, atx)
+	}
+	if len(val.activeset) == 0 {
+		val.before = append(val.before, a)
+	}
+
+	a.reg.register(val)
+	a.ballots[lid] = val
+	return val
+}
+
+type ballotAction struct {
+	types.BallotTortoiseData
+	base      *ballotAction
+	activeset []*atxAction
+	before    []action
+}
+
+func (b *ballotAction) String() string {
+	return fmt.Sprintf("ballot %v", b.ID)
+}
+
+func (b *ballotAction) deps() []action {
+	return b.before
+}
+
+func (b *ballotAction) computeOpinion() types.Hash32 {
+	votes := make([][]types.Vote, b.BallotTortoiseData.Layer.
+		Difference(types.GetEffectiveGenesis()),
+	) // vote on every layer before the ballot till effective genesis
+	for i := range votes {
+		votes[i] = []types.Vote{} // against by default
+	}
+	b.decodeVotes(votes)
+	hasher := opinionhash.New()
+	rst := types.Hash32{}
+	for i, layer := range votes {
+		if i != 0 {
+			hasher.WritePrevious(rst)
+		}
+		if layer == nil {
+			hasher.WriteAbstain()
+		}
+		if len(layer) > 0 {
+			sort.Slice(layer, func(i, j int) bool {
+				if layer[i].Height < layer[j].Height {
+					return true
+				}
+				return layer[i].ID.Compare(layer[j].ID)
+			})
+		}
+		for _, vote := range layer {
+			hasher.WriteSupport(vote.ID, vote.Height)
+		}
+		hasher.Sum(rst[:0])
+		hasher.Reset()
+	}
+	return rst
+}
+
+func (b *ballotAction) decodeVotes(votes [][]types.Vote) {
+	if b.base != nil {
+		b.base.decodeVotes(votes)
+	}
+	genesis := types.GetEffectiveGenesis()
+	for _, vote := range b.Opinion.Support {
+		pos := vote.LayerID.Difference(genesis)
+		matches := false
+		for _, existing := range votes[pos] {
+			if existing == vote {
+				matches = true
+			}
+		}
+		if !matches {
+			votes[pos] = append(votes[pos], vote)
+		}
+	}
+	for _, vote := range b.Opinion.Against {
+		pos := vote.LayerID.Difference(genesis)
+		for i, existing := range votes[pos] {
+			if existing == vote {
+				votes[pos] = append(votes[pos][:i], votes[pos][i:]...)
```

Comment:
do you mean `votes[pos] = append(votes[pos][:i], votes[pos][i+1:]...)`

Reply:
yes, thanks